File: node/antchain_node/node.py
"""Ties together chain, mempool, miner, and network into one running node."""
from __future__ import annotations

import socket
import threading
import time

from .block import Block
from .chain import Blockchain, ChainError
from .consensus import ConsensusParams
from .mempool import Mempool
from .miner import Miner
from .network import Network
from .transaction import Transaction


class Node:
    def __init__(self, host: str, port: int, params: ConsensusParams,
                 data_path: str | None, miner_address: str | None,
                 mine: bool, log=print):
        self.host = host
        self.port = port
        self.log = log
        self.chain = Blockchain(params, data_path=data_path)
        self.chain.load()
        self.mempool = Mempool()
        self.network = Network(host, port, self._on_message, log=log)
        self.miner_address = miner_address
        self.miner = (
            Miner(self.chain, self.mempool, miner_address, self._on_block_found, log=log)
            if mine and miner_address else None
        )

    # ------------------------------------------------------------- start
    def start(self, peers: list[tuple[str, int]]) -> None:
        self.network.start()
        for h, p in peers:
            if self.network.connect_to(h, p):
                self.network.send(self.network.peers[f"{h}:{p}"], {
                    "type": "hello", "host": self.host, "port": self.port, "height": self.chain.height,
                })
                self.network.send(self.network.peers[f"{h}:{p}"], {"type": "get_chain"})
        if self.miner is not None:
            self.miner.start()
            self.log(f"[node] mining enabled, rewards -> {self.miner_address}")

    # --------------------------------------------------------- callbacks
    def _on_block_found(self, block: Block) -> None:
        try:
            self.chain.try_extend(block)
            self.mempool.remove_all(block.transactions)
            self.network.broadcast({"type": "new_block", "block": block.to_dict()})
            self.log(f"[chain] height={self.chain.height} tip={self.chain.tip.hash[:16]}...")
        except ChainError as e:
            self.log(f"[miner] mined a block that failed self-validation: {e}")

    def _on_message(self, sock: socket.socket, msg: dict) -> None:
        t = msg.get("type")
        if t == "hello":
            return
        if t == "get_chain":
            self.network.send(sock, {"type": "chain", "blocks": self.chain.to_dict_list()})
        elif t == "chain":
            self._handle_full_chain(msg.get("blocks", []))
        elif t == "new_block":
            self._handle_new_block(msg.get("block"), origin=sock)
        elif t == "new_tx":
            self._handle_new_tx(msg.get("tx"), origin=sock)
        elif t == "get_balance":
            self.network.send(sock, {
                "type": "balance", "address": msg["address"],
                "balance": self.chain.balance_of(msg["address"]),
                "nonce": self.chain.next_nonce_for(msg["address"]),
            })
        elif t == "get_status":
            self.network.send(sock, {
                "type": "status", "height": self.chain.height, "tip": self.chain.tip.hash,
                "mempool_size": len(self.mempool), "peers": list(self.network.peers.keys()),
            })
        elif t == "submit_tx":
            try:
                tx = Transaction.from_dict(msg["tx"])
                ok = self.mempool.add(tx, self.chain)
                if ok:
                    self.network.broadcast({"type": "new_tx", "tx": tx.to_dict()}, exclude=sock)
                self.network.send(sock, {"type": "submit_tx_result", "accepted": ok, "tx_hash": tx.tx_hash})
            except Exception as e:
                self.network.send(sock, {"type": "submit_tx_result", "accepted": False, "error": repr(e)})

    def _handle_new_block(self, block_dict: dict | None, origin: socket.socket | None = None) -> None:
        if not block_dict:
            return
        block = Block.from_dict(block_dict)
        if block.hash == self.chain.tip.hash:
            return
        if block.prev_hash != self.chain.tip.hash:
            self.log(f"[chain] received block {block.index} that doesn't extend our tip; requesting full chain")
            self.network.broadcast({"type": "get_chain"})
            return
        try:
            self.chain.try_extend(block)
            self.mempool.remove_all(block.transactions)
            self.network.broadcast({"type": "new_block", "block": block.to_dict()}, exclude=origin)
            self.log(f"[chain] accepted block {block.index} from network, height={self.chain.height}")
        except ChainError as e:
            self.log(f"[chain] rejected block {block.index}: {e}")

    def _handle_full_chain(self, blocks_raw: list[dict]) -> None:
        if not blocks_raw:
            return
        try:
            blocks = [Block.from_dict(d) for d in blocks_raw]
        except (KeyError, TypeError):
            return
        try:
            if self.chain.try_replace(blocks):
                self.log(f"[chain] adopted longer/heavier chain, height={self.chain.height}")
        except ChainError as e:
            self.log(f"[chain] rejected candidate chain: {e}")

    def _handle_new_tx(self, tx_dict: dict | None, origin: socket.socket | None = None) -> None:
        if not tx_dict:
            return
        tx = Transaction.from_dict(tx_dict)
        if self.mempool.add(tx, self.chain):
            self.network.broadcast({"type": "new_tx", "tx": tx.to_dict()}, exclude=origin)

    def run_forever(self) -> None:
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            self.log("[node] shutting down")
            if self.miner is not None:
                self.miner.stop()

File: cpp_node/src/node.cpp
#include "node.hpp"

#include <chrono>
#include <stdexcept>
#include <thread>

namespace antchain_node {

namespace {
const Json kNullJson = Json::null();
}

Node::Node(std::string host, int port, ConsensusParams params, std::string data_path,
           std::string miner_addr, bool mine, Network::LogFn log)
    : chain(std::move(params), std::move(data_path)),
      network(std::move(host), port, [this](socket_t s, const Json& m) { on_message(s, m); }, log),
      miner_address(std::move(miner_addr)),
      log_(std::move(log)) {
    chain.load();
    if (mine && !miner_address.empty()) {
        miner = std::make_unique<Miner>(chain, mempool, miner_address,
                                         [this](const Block& b) { on_block_found(b); }, log_);
    }
}

void Node::log(const std::string& msg) const {
    if (log_) log_(msg); else std::puts(msg.c_str());
}

void Node::start(const std::vector<std::pair<std::string, int>>& peers) {
    network.start();
    for (const auto& hp : peers) {
        if (network.connect_to(hp.first, hp.second)) {
            socket_t s = network.peer_socket(hp.first + ":" + std::to_string(hp.second));
            if (s != kInvalidSocket) {
                Json hello = Json::object();
                hello.set("type", Json::string("hello"));
                hello.set("host", Json::string(network.host));
                hello.set("port", Json::integer(network.port));
                hello.set("height", Json::integer(static_cast<int64_t>(chain.height())));
                network.send(s, hello);
                Json get_chain = Json::object();
                get_chain.set("type", Json::string("get_chain"));
                network.send(s, get_chain);
            }
        }
    }
    if (miner) {
        miner->start();
        log("[node] mining enabled, rewards -> " + miner_address);
    }
}

void Node::on_block_found(const Block& block) {
    try {
        chain.try_extend(block);
        mempool.remove_all(block.transactions);
        Json msg = Json::object();
        msg.set("type", Json::string("new_block"));
        msg.set("block", block.to_json());
        network.broadcast(msg);
        log("[chain] height=" + std::to_string(chain.height()) + " tip=" + chain.tip().hash.substr(0, 16) + "...");
    } catch (const ChainError& e) {
        log(std::string("[miner] mined a block that failed self-validation: ") + e.what());
    }
}

void Node::handle_new_block(const Json* block_json, socket_t origin) {
    if (!block_json) return;
    Block block = Block::from_json(*block_json);
    if (block.hash == chain.tip().hash) return;
    if (block.prev_hash != chain.tip().hash) {
        log("[chain] received block " + std::to_string(block.index) + " that doesn't extend our tip; requesting full chain");
        Json get_chain = Json::object();
        get_chain.set("type", Json::string("get_chain"));
        network.broadcast(get_chain);
        return;
    }
    try {
        chain.try_extend(block);
        mempool.remove_all(block.transactions);
        Json msg = Json::object();
        msg.set("type", Json::string("new_block"));
        msg.set("block", block.to_json());
        network.broadcast(msg, origin);
        log("[chain] accepted block " + std::to_string(block.index) + " from network, height=" + std::to_string(chain.height()));
    } catch (const ChainError& e) {
        log("[chain] rejected block " + std::to_string(block.index) + ": " + e.what());
    }
}

void Node::handle_full_chain(const Json& blocks_json) {
    if (!blocks_json.is_array() || blocks_json.size() == 0) return;
    std::vector<Block> blocks;
    try {
        for (const auto& bj : blocks_json.as_array()) blocks.push_back(Block::from_json(bj));
    } catch (const std::exception&) {
        return;
    }
    try {
        if (chain.try_replace(blocks)) log("[chain] adopted longer/heavier chain, height=" + std::to_string(chain.height()));
    } catch (const ChainError& e) {
        log(std::string("[chain] rejected candidate chain: ") + e.what());
    }
}

void Node::handle_new_tx(const Json* tx_json, socket_t origin) {
    if (!tx_json) return;
    Transaction tx = Transaction::from_json(*tx_json);
    if (mempool.add(tx, chain)) {
        Json msg = Json::object();
        msg.set("type", Json::string("new_tx"));
        msg.set("tx", tx.to_json());
        network.broadcast(msg, origin);
    }
}

void Node::on_message(socket_t sock, const Json& msg) {
    if (!msg.is_object() || !msg.has("type")) return;
    std::string type = msg.at("type").as_string();

    if (type == "hello") return;

    if (type == "get_chain") {
        Json reply = Json::object();
        reply.set("type", Json::string("chain"));
        reply.set("blocks", chain.to_json_list());
        network.send(sock, reply);
    } else if (type == "chain") {
        handle_full_chain(msg.get("blocks", kNullJson));
    } else if (type == "new_block") {
        if (msg.has("block")) { const Json& b = msg.at("block"); handle_new_block(&b, sock); }
    } else if (type == "new_tx") {
        if (msg.has("tx")) { const Json& t = msg.at("tx"); handle_new_tx(&t, sock); }
    } else if (type == "get_balance") {
        std::string addr = msg.at("address").as_string();
        Json reply = Json::object();
        reply.set("type", Json::string("balance"));
        reply.set("address", Json::string(addr));
        reply.set("balance", Json::number(chain.balance_of(addr)));
        reply.set("nonce", Json::integer(static_cast<int64_t>(chain.next_nonce_for(addr))));
        network.send(sock, reply);
    } else if (type == "get_status") {
        Json reply = Json::object();
        reply.set("type", Json::string("status"));
        reply.set("height", Json::integer(static_cast<int64_t>(chain.height())));
        reply.set("tip", Json::string(chain.tip().hash));
        reply.set("mempool_size", Json::integer(static_cast<int64_t>(mempool.size())));
        Json peer_list = Json::array();
        for (const auto& kv : network.peers()) peer_list.push_back(Json::string(kv.first));
        reply.set("peers", peer_list);
        network.send(sock, reply);
    } else if (type == "submit_tx") {
        Transaction tx;
        bool parsed = false;
        std::string error;
        try {
            tx = Transaction::from_json(msg.at("tx"));
            parsed = true;
        } catch (const std::exception& e) {
            error = e.what();
        }
        Json reply = Json::object();
        reply.set("type", Json::string("submit_tx_result"));
        if (!parsed) {
            reply.set("accepted", Json::boolean(false));
            reply.set("error", Json::string(error));
        } else {
            bool ok = mempool.add(tx, chain);
            if (ok) {
                Json bcast = Json::object();
                bcast.set("type", Json::string("new_tx"));
                bcast.set("tx", tx.to_json());
                network.broadcast(bcast, sock);
            }
            reply.set("accepted", Json::boolean(ok));
            reply.set("tx_hash", Json::string(tx.tx_hash()));
        }
        network.send(sock, reply);
    }
}

void Node::run_forever() {
    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

} // namespace antchain_node
